File: src/codeforces/contest1731/ProblemD.java
package codeforces.contest1731;

import java.util.Scanner;

public class ProblemD {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int tests = scanner.nextInt();
        for (int t = 0; t < tests; t++) {
            int m = scanner.nextInt();
            int n = scanner.nextInt();
            int[][] a = new int[m][n];
            for (int i = 0; i < m; i++) for (int j = 0; j < n; j++) a[i][j] = scanner.nextInt();
            System.out.println(solve(a));
        }
    }

    private static int solve(int[][] a) {
        int m = a.length;
        int n = a[0].length;
        int lo = 1, hi = Math.min(m, n);
        while (lo < hi) {
            int mid = lo + (hi - lo + 1) / 2;
            if (check(a, m, n, mid)) lo = mid;
            else hi = mid - 1;
        }
        return lo;
    }

    private static boolean check(int[][] a, int m, int n, int val) {
        int[] prev = new int[n + 1], next = new int[n + 1], tmp;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (a[i][j] >= val) {
                    next[j + 1] = Math.min(prev[j], Math.min(next[j], prev[j + 1])) + 1;
                    if (next[j + 1] >= val) return true;
                } else next[j + 1] = 0;
            }
            tmp = prev;
            prev = next;
            next = tmp;
        }
        return false;
    }

}

File: src/codeforces/contest1676/ProblemD.kt
@file:Suppress("DuplicatedCode")

package codeforces.contest1676

import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.util.*
import kotlin.math.max

fun main(args: Array<String>) {
    val tests = nextInt()
    for (t in 0 until tests) {
        val n = nextInt()
        val m = nextInt()
        val arr = Array(n) { nextInts(m) }
        println(solve(arr))
    }
}

fun solve(arr: Array<IntArray>): Long {
    val m = arr.size
    val n = arr[0].size

    // x+y = (0, m+n-2)
    val d1 = LongArray(m + n - 1)
    // x-y + n-1  = (0, m+n-2)
    val d2 = LongArray(m + n - 1)
    for (i in 0 until m) {
        for (j in 0 until n) {
            d1[i + j] = d1[i + j] + arr[i][j]
            d2[i - j + n - 1] = d2[i - j + n - 1] + arr[i][j]
        }
    }

    var ans = Long.MIN_VALUE
    for (i in 0 until m) {
        for (j in 0 until n) {
            ans = max(ans, d1[i + j] + d2[i - j + n - 1] - arr[i][j])
        }
    }
    return ans
}


private class MyReader(inputStream: InputStream) {
    private val reader = BufferedReader(InputStreamReader(inputStream))
    private var tokenizer: StringTokenizer? = null

    fun next(): String {
        while (true) {
            tokenizer.let {
                if (it == null || !it.hasMoreTokens()) {
                    tokenizer = StringTokenizer(reader.readLine())
                } else {
                    return it.nextToken()
                }
            }
        }
    }
}

private val reader = MyReader(System.`in`)
private fun next() = reader.next()
private fun nextInt() = next().toInt()
private fun nextLong() = next().toLong()
private fun nextDouble() = next().toDouble()
private fun nextStrings(n: Int) = Array(n) { next() }
private fun nextInts(n: Int) = IntArray(n) { nextInt() }
private fun nextLongs(n: Int) = LongArray(n) { nextLong() }
private fun nextDoubles(n: Int) = DoubleArray(n) { nextDouble() }